File: templates/docs-structure/examples/python/perfect_service.py
"""EXAMPLE — canonical pattern, copy & adapt.

A service: pure business logic, framework-agnostic.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from perfect_repository import UserRepository


class NotFoundError(Exception):
    pass


@dataclass(frozen=True)
class User:
    id: str
    email: str


class UserService:
    def __init__(self, repo: UserRepository) -> None:
        self._repo = repo

    def get_user(self, user_id: str) -> User:
        user = self._repo.find_by_id(user_id)
        if user is None:
            raise NotFoundError(user_id)
        return user

File: templates/docs-structure/examples/node/perfect-service.ts
// EXAMPLE — canonical pattern, copy & adapt.
// A service: pure business logic, no framework/transport concerns.
export interface UserRepo {
  findById(id: string): Promise<{ id: string } | null>;
}

export class UserService {
  constructor(private readonly repo: UserRepo) {}

  async getUser(id: string) {
    const user = await this.repo.findById(id);
    if (!user) throw new Error('NotFound');
    return user;
  }
}
